Support JSON Lines importing/exporting

WiredSound opened the pull request:
This pull request hopes to resolve this issue: [Support JSON Lines format (.jsonl) #183](https://github.com/getsynth/synth/issues/183)

This is PR adds support from importing and exporting data as JSON Lines. It also introduces a URI system for specifying input/output data types using the `--from` and `--to` CLI arguments.

## Usage

Using a URI schema `jsonl:` (specified with `generate --to` or `import --from`) users can now output/input JSON Lines data. Here's an example using the `bank_db` example schema:

```
$ synth generate examples/bank/bank_db --to jsonl:
{"type":"users","created_at":"2015-08-13T05:08:41+0000","credit_card":"342044649251781","email":"margarete@example.org","id":1,"is_active":false,"last_login_at":"2019-07-31T05:16:20+0000","num_logins":24,"username":"margarete@example.org"}
{"type":"users","created_at":"2015-01-18T01:20:40+0000","credit_card":"4564426869098","email":"rebeka@example.com","id":2,"is_active":false,"last_login_at":"2018-07-06T02:40:49+0000","num_logins":85,"username":"art_tenetur"}
{"tyoe":"users","created_at":"2015-03-09T16:45:35+0000","credit_card":"4640858388993","email":"aida@example.net","id":3,"is_active":false,"last_login_at":"2018-02-03T00:59:04+0000","num_logins":75,"username":"aida@example.net"}
{"amount":4346.4,"type":"transactions","currency":"USD","id":1000,"timestamp":"2015-09-29T06:22:09+0000","user_id":1}
{"amount":700.8,"type":"transactions","currency":"GBP","id":1001,"timestamp":"2008-05-04T03:19:15+0000","user_id":2}
{"amount":1243.5,"type":"transactions","currency":"USD","id":1002,"timestamp":"2005-07-31T06:28:24+0000","user_id":3}
```

Note the `:` in URI `jsonl:` - this is a valid URI, just without a path specified. URIs without a path specified will use stdin/stdout in the data format specified by the scheme (i.e. `json:` or `jsonl:`). Also note the `"type"` field that specifies what collection the JSONL object was generated with. You can change the name of this field by adding an argument to the URI string `collection_field_name`:

```
$ synth generate example/bank/bank_db --to jsonl:?collection_field_name=abc
{"abc":"users","created_at":"2015-08-13T05:08:41+0000","credit_card":"342044649251781","email":"margarete@example.org","id":1,"is_active":false,"last_login_at":"2019-07-31T05:16:20+0000","num_logins":24,"username":"margarete@example.org"}}
...
```

Specifying an individual collection to work with via `--collection` will mean that such a field won't be inserted into the generated objects at all:

```
$ synth generate examples/bank/bank_db --to jsonl: --collection transactions
{"amount":4346.4,"currency":"USD","id":1000,"timestamp":"2015-09-29T06:22:09+0000","user_id":1}
...
```

You can of course import JSON Lines data into Synth as well. Assuming a file `data.jsonl` which contains the output of the first command demonstrated in this section:

```
$ synth import namespace --from jsonl:data.jsonl

$ ls namespace/
transactions.json  users.json

$ cat namespace/transactions.json
{
  "type": "array",
  "length": {
    "type": "number",
    "range": {
      "low": 1,
      "high": 3,
      "step": 1
    },
    "subtype": "u64"
  },
  "content": {
    "type": "object",
    "amount": {
      "type": "number",
      "range": {
        "low": 700.8,
        "high": 4347.4,
        "step": 1.0
      },
      "subtype": "f64"
    },
    "currency": {
      "type": "string",
      "pattern": "[a-zA-Z0-9]*"
    },
...
```

Note the lack of a `//` you may have expected - this is because there will not be an authority component as part of the URI when just interacting with the local filesystem.

## Problems

What should Synth do with schemas that do not have `"type": "object"` for the outer most `"content"` of the schema (i.e. a schema that is generating an array of something other than objects)?

```
{
        "type": "array",
        "length": {
                "type": "number",
                "range": { "low": 1, "high": 3, "step": 1 },
                "subtype": "u64"
        },
        "content": {
                "type": "array",
                "length": {
                        "type": "number",
                        "constant": 2,
                        "subtype": "u64"
                },
                "content": {
                        "type": "string",
                        "faker": {
                                "generator": "address"
                        }
                }
        }
}
```

With the above schema, running a command like `synth generate namespace --to jsonl:output.jsonl --collection foobar` would generate:

```
["56631 Arjun Ways, AR 169","93701 Brenda Glen, PA 38630"]
["1214 Casper Forges, NJ 202","80556 Douglas Mill, NH 465"]
```

However if no `--collection` argument is specified, currently that will produce output like the following:

```
{"collection":"foobar","data":["56631 Arjun Ways, AR 169","93701 Brenda Glen, PA 38630"]}
{"collection":"foobar","data":["1214 Casper Forges, NJ 202","80556 Douglas Mill, NH 465"]}
```

In order to allow for multiple data from multiple collections to be generated in a namespace, the output array is embedded in an object which specifies the collection. This seems problematic as I can see no good way for Synth to properly import such data - how would it know if the user had originally generated an object with a `"data"` field or just the array? Opinions on this would be appreciated. At the moment I feel it may make sense to just prevent users from generating non-object types when dealing with JSONL as I can't really imagine a use case for doing so anyway.

brokad reviewed (commented):
On the whole, LGTM, thanks @WiredSound! 😄 

Just a few (obvious) left-overs in the comments. Also, it'd be nice to add a quick and dirt smoke test for the `--to/--from jsonl:` stuff.

WiredSound commented on docs/docs/examples/bank.md:
````diff
@@ -52,7 +52,7 @@ mkdir synth_workspace && cd synth_workspace && synth init
 
 ```bash
-synth import bank_db/ --from /path/to/the/file.json
+synth import bank_db/ --from json:/path/to/the/file.json
````
It is a necessary change as a path just on it's own isn't a valid URI and as it stands `--from` now only accepts URIs. Perhaps it would be worth adding support for specifying just file paths with the assumption that a path deals with regular JSON?

WiredSound commented on docs/docs/getting_started/cli.md:
```diff
@@ -21,18 +21,18 @@ If a subdirectory for a given namespace does not exist, Synth will create it.
+
+  Importing from a database is possible using the standard URI format of the respective database. For example: `postgres://user:pass@localhost:5432/tpch`
+
+  When dealing with JSON Lines and not specifying a single collection with the `--collection` argument, each generated object is tagged with the name of the collection it was generated from. By default, this is done by adding a property `type` to the object (e.g. `"type": "collection_name"`). The name of this property can be changed using an additional parameter `collection_field_name` added at t ...
```
It was indeed! I haven't used CouchDB before but based on a quick Google search it looks relatively straightforward to write a shell script to import JSON Lines into a database - could likely work well in CI.

hbina replied:
Is it necessary to support the `collection_field_name=foobar` query? I prefer leaning into more UNIXy solution to this. Why not just support pipe input and utilize something like `jq` for this kind of data manipulation instead? It is also way more powerful than this and avoids Not Invented Here problem

WiredSound replied:
@hbina To tell the truth, I would be in favour of relying more heavily of Unix pipes, etc. in general but enough users are on Windows where such operations are awkward that we decided against it. Also specifically with regards to `collection_field_name`, I think it would be useful to set that in advance (rather than modify it afterwards with `jq`) as you may for example have a schema that produces an object with a property `"type"`, in which case you probably don't want Synth overwriting that data, or silently renaming the property's key, or otherwise.

WiredSound commented on synth/src/cli/export.rs:
```diff
@@ -1,60 +1,76 @@
-                    ));
-                };
-                Ok(export_strategy)
+                    unimplemented!();
```
Yeah, I wasn't sure if it would fall within the scope of this PR as exporting file seems to have been unimplemented for a while - though I suppose it would take very little to implement, I may as well do it now!

WiredSound commented on synth/src/cli/export.rs:
```diff
@@ -1,60 +1,76 @@
+        // 'mysql' suprisingly) it seems simpler to just match based on the scheme string instead of on enum variants.
+        let scheme = params.uri.scheme().as_str().to_lowercase();
+        let export_strategy: Box<dyn ExportStrategy> = match scheme.as_str() {
+            "postgres" | "postgresql" => Box::new(PostgresExportStrategy {
```
I don't believe this should be a problem as ultimately the PostgreSQL database driver handles this - Synth only has to match on the URI scheme and external libraries should handle the rest. I tried a couple inputs and `postgres://localhost/my_db` and `postgres:my_db` and `postgres:///my_db` are essentially equivalent which makes sense to me based on my understanding - RFC 3936 defines an 'authority' as `[ userinfo "@" ] host [ ":" port ]` so in this example `localhost` is the authority component. You can *optionally* give the authority component when `//` is used. For `postgres://localhost/my_db` the authority is given explicitly; for `postgres:my_db` Postgres assumes a default authority; and for `postgres:///my_db`, despite the `//`, the authority doesn't actually need to be given so again a default is assumed.

WiredSound commented:
Thank you for the feedback! @brokad 
I'll write some general JSONL in/out tests and probably implement generation to file tomorrow.

christoshadjiaslanis commented:
Thanks @WiredSound - this looks great :) Just a few things I think we can improve.

0. Your PR has made it clear to me that we are conflating two ideas here - the idea of an export/import strategy and a data source (csv, jsonl, and json can output to a file - yet the way they are serialized into from a `synth::Value` is different). This is outside the scope of this PR but I thought I would mention it for the record.
1. Everything on the export side looks great. I would split the  export strategy into two however - have one for `json` and one for `jsonl`. Also you can bring in the logic added to the `sampler::into_json_lines` to the `JsonlStdoutExportStrategy` (and at a later date move other logive from the sampler to export strategies as well)
2. I would say the same for import strategy. I think we should remove the default implementation of `ImportStrategy::import` and make it an associated function (implement it on a per-strategy basis). This would mean creating separate strategies for `jsonl` and `json` and should remove the functions `get_data_format`, `as_json_value` etc. Due to point 0, this may result in 4 strategies for now but it improves clarity and this is tech debt we can resolve at a later date. This should also remove the `DataFormat` enum. 

I think we're nearly there! Lmk if there are any questions or if I have missed something (which may very well be the case) :)

WiredSound commented:
Thanks @christoshadjiaslanis - think I've now addressed the points you've raised :) JSON and JSONL import/exporting are now further separated (`JsonFileExportStrategy`, `JsonLinesFileExportStrategy`, `JsonStdinImportStrategy`, `JsonLinesStdinImportStrategy`) and things have been reorganised somewhat.

christoshadjiaslanis commented:
Thank you @WiredSound . Great work :)
